### DISCUSSION TITLE
Handling large amounts of static, CMS-based (JSON) data

### DISCUSSION THREAD

**Discussion by nathnhughes** [vercel/next.js#37065] [Help]
# The problem
I recently worked on a site that fetched articles from a Contentful CMS. The articles were written with rich text and were usually fairly substantial, however, they were totally static meaning once they'd been written they wouldn't change.

As the data was static, we wanted to fetch it at build time using `getStaticProps` and pre-generate the HTML for articles.

Now, the way Contentful formats rich text is in the form of an _(often gigantic)_ JSON object. Here's an example: 

```json
{
  "content": {
    "json": {
      "nodeType": "document",
      "data": {},
      "content": [
        {
          "nodeType": "paragraph",
          "data": {},
          "content": [
            {
              "nodeType": "text",
              "value": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. ",
              "marks": [],
              "data": {}
            },
            {
              "nodeType": "entry-hyperlink",
              "data": {
                "target": {
                  "sys": { "id": "5TDMISmspS0flb7xXy3ifO", "type": "Link", "linkType": "Entry" }
                }
              },
              "content": [{ "nodeType": "text", "value": "Lorem", "marks": [], "data": {} }]
            },
            {
              "nodeType": "text",
              "value": "ipsum dolor sit amet, consectetur adipiscing elit.",
              "marks": [],
              "data": {}
            }
          ]
        },
        {
          "nodeType": "paragraph",
          "data": {},
          "content": [
            {
              "nodeType": "text",
              "value": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
              "marks": [],
              "data": {}
            }
          ]
        },
        {
          "nodeType": "heading-2",
          "data": {},
          "content": [{ "nodeType": "text", "value": "Lorem", "marks": [], "data": {} }]
        },
        {
          "nodeType": "paragraph",
          "data": {},
          "content": [
            {
              "nodeType": "text",
              "value": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt",
              "marks": [],
              "data": {}
            },
            { "nodeType": "text", "value": "Lorem ", "marks": [{ "type": "italic" }], "data": {} },
            {
              "nodeType": "text",
              "value": " consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "marks": [],
              "data": {}
            },
            { "nodeType": "text", "value": "lorem ", "marks": [{ "type": "italic" }], "data": {} },
            { "nodeType": "text", "value": "(pisum) dolor ", "marks": [], "data": {} },
            { "nodeType": "text", "value": "amet ", "marks": [{ "type": "italic" }], "data": {} },
            {
              "nodeType": "text",
              "value": "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae",
              "marks": [],
              "data": {}
            },
            {
              "nodeType": "text",
              "value": "cillum dolore",
              "marks": [{ "type": "italic" }],
              "data": {}
            },
            { "nodeType": "text", "value": ").”", "marks": [], "data": {} },
            {
              "nodeType": "hyperlink",
              "data": { "uri": "#lorem" },
              "content": [{ "nodeType": "text", "value": "sint", "marks": [], "data": {} }]
            },
            { "nodeType": "text", "value": "", "marks": [], "data": {} }
          ]
        },
        {
          "nodeType": "heading-3",
          "data": {},
          "content": [{ "nodeType": "text", "value": "Impsum", "marks": [], "data": {} }]
        },
        {
          "nodeType": "embedded-entry-block",
          "data": {
            "target": { "sys": { "id": "1LI8rNPmyo7ZDKs0v9UQn6", "type": "Link", "linkType": "Entry" } }
          },
          "content": []
        }
      ]
    }
  }
}
```

This data is around 100 lines worth of an object that could often be 3500+ lines _(when prettified)_ long.

Seeing as Next has to provide these props on the server and rehydrate the static markup on the client, this rich text JSON would be included in the `__NEXT_DATA__` script and we'd regularly get [`large-page-data`](https://nextjs.org/docs/messages/large-page-data) warning from the server logs reporting page data of 150kb+. 

<br>

# Question
Wondering whether anyone has run into a similar situation and how they've handled it? To my mind, there must be a more efficient way of doing this but until now I haven't been able to find one.

As the content is static, and for performance and SEO benefits, I'm keen to use `gSP` to fetch the data at build time and pre-render the HTML for the server. However, I feel that the massive size of the data that has to be included in `__NEXT_DATA__` may actually be having the opposite effect. 

When looking in the .next folder after running a local build, you can see the server-side HTML pages with all the data already there, so does it need to be passed to the client to 'rehydrate' when it shouldn't have changed?

<br>

# One other point
One other thing I have noticed when looking around at other Next sites doing a similar thing is that using markdown seems to yield much better performance. This is most likely down to the much more condensed syntax and the fact that the entire article is sent as a single string.

For example, to generate the following sentence: 

"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt _consectetur_ adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."

using JSON rich-text, you would need:

```json
{
  "nodeType": "paragraph",
  "data": {},
  "content": [
    {
      "nodeType": "text",
      "value": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt",
      "marks": [],
      "data": {}
    },
    { "nodeType": "text", "value": "consectetur ", "marks": [{ "type": "italic" }], "data": {} },
    {
      "nodeType": "text",
      "value": " adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
      "marks": [],
      "data": {}
    }
  ]
}
```

using markdown, you would need:

```json
{
  "data": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt _consectetur_ adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
}
```

With this in mind, I've actually even looked into transforming the Contentful JSON response in markdown and then passing it to the component, but this seems pretty convoluted as once it's received by the component, it's just transformed again using `@contenful/rich-text-react-renderer`, so we're transforming the data twice...

Would love to hear if people are facing the same issue or if anyone has any bright ideas!

**Comment by wwwdepot** [comment]
Use `@contentful/rich-text-html-renderer`, transform CMS document to HTML string inside `getStaticProps` & pass the string to the page component so you can dangerouslySetInnerHTML it.

**Reply by nathnhughes** [reply]
Yep, have considered this, however,  `@contenful/rich-text-react-renderer` is being used to transform the rich-text data to specific React components rather than plain HTML.

The rich text also contains things like this: 

```json
{
  "nodeType": "embedded-entry-block",
  "data": {
    "target": { 
      "sys": { 
        "id": "1LI8rNPmyo7ZDKs0v9UQn6", 
        "type": "Link",
        "linkType": "Entry" 
      } 
    }
  },
  "content": []
}
```

which are links to other entries in the CMS _(for example, a `mediaAsset` like an image that's also stored in the CMS)_. 

These links need to be interpreted by some sort of component otherwise they just render as span tags.

**Reply by wwwdepot** [reply]
You can pass the `options` in the 2nd parameter to customize how to render those embedded entry. 

Anyway, if you need it to be interactive, you want to stick it with `@contenful/rich-text-react-renderer` to use `next/link` & `next/image` or other React components.
